keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+- The k/kubeadm tags and branches need to follow those of k/k
+- k/k tags are not SEMVER, which can cause troubles when importing
+the k/kubeadm repository
+
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Building kubeadm from tarballs
+
+For the initial proposal this will not be supported, but if there
+are plans to distribute the kubeadm source code as part of a k8s release,
+the solution would be to write a static KUBE_VERSION file in the tarball.
+
+If GitHub users download the source code directly from the kubeadm repository
+the build process must be able to detect that this is not a Git repository
+and ask the users to provide a KUBE_VERSION file.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
+
+##### Syncing tags and branches between k/k and k/kubeadm
+
+A brief discussion of the #prow maintainers revealed the proposal that
+a post-submit job can be used for that instead of prow plugins.
+
+Supposedly such a job will trigger each time a new branch is created
+or if a tag is pushed. This means that the post-submit can trigger any bash
+script or application to get the latest list of branches and tags in k/k
+and apply the same to the kubeadm repository.
+
+Given kubeadm will potentially move within a single cycle this means that
+not all tags and branches have to be synced - see the options in “Versions
+in the support skew”. So potentially there must be a minimum version
+check in the post-submit job.
+
+Scenarios:
+
+###### k/k creates a new release-x.yy branch
+
+The post-submit job checks out master in k/kubeadm, creates a new branch
+release-x.yy and pushes the branch.
+
+###### k/k ‘s release-x.yy branch is fast-forwarded to master
+
+Currently this is a manual step performed by someone on the release team.
+It might be something that the k/kubeadm maintainers have to do manually
+on a daily basis. Automation is possible but is tricky and needs further
+discussion. Can be done as a nice-to-have in the future.
```

Comment:
This would only apply during the beta development phase.

That still leaves the rc and official release phases where tags are applied, and well as patch release phase.  We need a process/mechanism that ensures that k/kubeadm branches are in a release ready state ahead of k/k's "go" decision on an official .0 or subsequent patch release.

Reply:
is "beta development phase" the period after the branch is created and before the first RC?

> We need a process/mechanism that ensures that k/kubeadm branches are in a release ready state ahead of k/k's "go" decision on an official .0 or subsequent patch release.

i think this process can be looking at the periodic e2e test signal for kubeadm.
this can be automated in the future for both k/k and k/kubeadm - i.e. the release manager gett a notification that something is really broken based on e2e test signal.

currently how does the branch manager knows if k/k is ready for a new release? do they check all e2e tests for the kubelet and control-plane components?

Reply:
Yes manual fast forwarding is happening after beta and before rc.

Branch manager is managing content and readiness relative to sig-release-X.YY-blocking and informing boards, CI signal interpretation of those boards, and also developer interactive discussion (eg: we regularly have scenarios of XX must merge before YY or in some specific ordering and conditioned on some other status also).

Reply:
i don't see the potential for differences between the release team process for k/k/kubeadm and k/kubeadm. the source of truth is currently test-signal. this will continue to happen to be the case.

the release-team will be notified on __major__ blocking issues on the side of k/kubeadm side.
e.g. someone deleted a branch by mistake.

Reply:
Really the summary in the KEP here is already summarizing this thread, ie: "Automation is possible but is tricky and needs further discussion."